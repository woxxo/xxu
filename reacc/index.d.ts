type ReactNode = any;

export interface RenderToReadableStreamOptions {
	identifierPrefix?: string;
	namespaceURI?: string;
	nonce?: string;
	bootstrapScriptContent?: string;
	bootstrapScripts?: string[];
	bootstrapModules?: string[];
	progressiveChunkSize?: number;
	signal?: AbortSignal;
	onError?: (error: unknown, errorInfo: unknown) => string | void;
}

export interface ReactDOMServerReadableStream extends ReadableStream {
	allReady: Promise<void>;
}

export function renderToReadableStream(
	children: ReactNode,
	options?: RenderToReadableStreamOptions,
): Promise<ReactDOMServerReadableStream>;

