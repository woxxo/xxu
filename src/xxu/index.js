export class Xxu {
	static #VERSION = 'XXU server v.0.4.0';
	
	static #defaultHandler = ({ method, headers, body, path, params }) => {
		return `xxu: ${method}\nxxu: ${path}`;
	}

	static #isBunFile = (file) => {
		return file instanceof Blob && 'name' in file && file.name.length > 0;
	};

	#server;
	#port = 3050;
	#hostname = '127.0.0.1';  //for docker 0.0.0.0
	#handler = Xxu.#defaultHandler;
	#serverSettings = {
		port: this.#port,
		hostname: this.#hostname,
		development: false,
		id: `${Xxu.#VERSION} #${new Date().getMilliseconds()}`,
		fetch: async ({method, url, headers, body}) => {
			const req = {};  //collecting request data for handler

			req.method = method;
			req.headers = headers;
			switch (method) {
				case 'POST':
				case 'DELETE':
				case 'PUT':
				case 'PATCH':
					req.body = body; //readable stream
					break;
				default:
					req.body = null;
			}

			const pathStart = url.indexOf('/', 11);
			const pathEnd = url.indexOf('?', pathStart + 1);
			const path = url.substring(pathStart, pathEnd >>> 0);
			const params = url.substring(pathEnd >>> 0);

			req.path = path;
			req.params = params;

			let resp;

			switch (this.#handler?.constructor.name) {
				case 'Function':
					resp = this.#handler(req);
					break;
				case 'AsyncFunction':
					resp = await this.#handler(req);
					break;
				default:
					console.log('Handler is not a function!');
					resp = new Response('Wrong handler!', { status: 500 });
					setImmediate(this.stop.bind(this));
			}

			if (Xxu.#isBunFile(resp)) {
				return new Response(resp);  //file
			}

			if (resp?.constructor.name === 'Response') {
				return resp; //response prepared completely by user
			}

			return new Response(resp, {
				status: 200,
				headers: {
					'Content-Type': 'text/html; charset=utf-8',
					//'Set-Cookie': `session${(Math.random() * 9999999 | 0)}=${(Math.random() * 9999999 | 0)}; SameSite=Lax; path=/`,
				 },
			});
		},
	};

	constructor(port = this.#port, handler = this.#handler) {
		this.#serverSettings.port = port;
		this.#handler = handler;
		this.#server = Bun.serve(this.#serverSettings);
		console.log(`Server ${this.#server.id} was started.`);
	}

	get version() {
		return Xxu.#VERSION;
	}

	get port() {
		return this.#server.port;
	}

	get hostname() {
		return this.#server.hostname;
	}

	stop() {
		console.log(`Server ${this.#server.id} was stopped.`);
		this.#server.stop(true); //forced to close all connections
	}

	setHandler(handler = Xxu.#defaultHandler) {
		this.#handler = handler;
	}

}
