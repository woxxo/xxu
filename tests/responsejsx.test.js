import { expect, test } from "bun:test";
import { Xxu } from 'xxu';
import { renderToReadableStream } from 'xxu/reacc';

const xxu = new Xxu(3091, async (method, path) => {
	switch(path) {
		case '/simple':
			return await renderToReadableStream(<div>Hello, world!</div>);
		case '/full':
			return await renderToReadableStream(
				<html lang="en">
				<head>
					<meta charset="utf-8" />
					<meta name="description" content="Default description" />
					<title>Default</title>
					<style>{ /* css */ `body {font-family: sans-serif; line-height: 1.5;}`}</style>
				</head>
				<body>
					{ [1, 2, 3, 4, 5].map(_ => <p className="myclass" key={_} style={{ padding: `${_}px` }}>N{_}N</p>) }
				</body>
				</html>, {
				bootstrapScriptContent: 'alert(1);',
				bootstrapScripts: ['/main1.js', '/main2.js', ],
			});
		default:
			return 'OK';
	}
});

test("response simple", async () => {
	let response = '';
	try {
		response = await fetch("http://127.0.0.1:3091/simple");
	} catch(e) {}
	expect(await response.text()).toEqual(`<div>Hello, world!</div>`);
});

test("response full", async () => {
	let response = '';
	try {
		response = await fetch("http://127.0.0.1:3091/full");
	} catch(e) {}
	expect(await response.text()).toMatchSnapshot();
});

test("server stopped", async () => {
	let response = 'stopped!';
	xxu.stop();
	try {
		response = await fetch("http://127.0.0.1:3091");
	} catch(e) {}
	expect(response).toEqual('stopped!');
});
