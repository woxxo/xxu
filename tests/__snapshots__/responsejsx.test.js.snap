// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`response full 1`] = `"<!DOCTYPE html><html lang="en"><head><link rel="preload" as="script" fetchPriority="low" href="/main1.js"/><link rel="preload" as="script" fetchPriority="low" href="/main2.js"/><meta charset="utf-8"/><meta name="description" content="Default description"/><title>Default</title><style>body {font-family: sans-serif; line-height: 1.5;}</style></head><body><p class="myclass" style="padding:1px">N<!-- -->1<!-- -->N</p><p class="myclass" style="padding:2px">N<!-- -->2<!-- -->N</p><p class="myclass" style="padding:3px">N<!-- -->3<!-- -->N</p><p class="myclass" style="padding:4px">N<!-- -->4<!-- -->N</p><p class="myclass" style="padding:5px">N<!-- -->5<!-- -->N</p><script>alert(1);</script><script src="/main1.js" async=""></script><script src="/main2.js" async=""></script></body></html>"`;
